extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergKeyValuePartitionedLayout.java

```diff
@@ -0,0 +1,92 @@
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+/**
+ * Iceberg {@link TableLocationKeyFinder location finder} for tables without partitions that will discover data files
+ * from a {@link org.apache.iceberg.Snapshot}
+ */
+public final class IcebergKeyValuePartitionedLayout implements TableLocationKeyFinder<IcebergTableLocationKey> {
+    private final Snapshot tableSnapshot;
+    private final FileIO fileIO;
+    private final String[] partitionColumns;
+    private final Object readInstructions;
+    private final Map<URI, IcebergTableLocationKey> cache;
+
+    private static IcebergTableLocationKey locationKey(
+            final FileFormat format,
+            final URI fileUri,
+            final Map<String, Comparable<?>> partitions,
+            @NotNull final Object readInstructions) {
+        return new IcebergTableLocationKey(format, fileUri, 0, partitions, readInstructions);
+    }
+
+    /**
+     * @param tableSnapshot The {@link Snapshot} from which to discover data files.
+     * @param fileIO The file IO to use for reading manifest data files.
+     * @param partitionColumns The columns to use for partitioning.
+     * @param readInstructions The instructions for customizations while reading.
+     */
+    public IcebergKeyValuePartitionedLayout(
+            @NotNull final Snapshot tableSnapshot,
+            @NotNull final FileIO fileIO,
+            @NotNull final String[] partitionColumns,
+            @NotNull final Object readInstructions) {
+        this.tableSnapshot = tableSnapshot;
+        this.fileIO = fileIO;
+        this.partitionColumns = partitionColumns;
+        this.readInstructions = readInstructions;
+
+        this.cache = new HashMap<>();
+    }
+
+    public String toString() {
+        return IcebergFlatLayout.class.getSimpleName() + '[' + tableSnapshot + ']';
+    }
+
+    @Override
+    public synchronized void findKeys(@NotNull final Consumer<IcebergTableLocationKey> locationKeyObserver) {
+        final Map<String, Comparable<?>> partitions = new LinkedHashMap<>();
+        try {
+            // Retrieve the manifest files from the snapshot
+            final List<ManifestFile> manifestFiles = tableSnapshot.allManifests(fileIO);
+            for (final ManifestFile manifestFile : manifestFiles) {
+                final ManifestReader<DataFile> reader = ManifestFiles.read(manifestFile, fileIO);
+                for (DataFile df : reader) {
+                    final URI fileUri = FileUtils.convertToURI(df.path().toString(), false);
+                    IcebergTableLocationKey locationKey = cache.get(fileUri);
+                    if (locationKey == null) {
+                        final PartitionData partitionData = (PartitionData) df.partition();
+                        for (int ii = 0; ii < partitionColumns.length; ++ii) {
+                            partitions.put(partitionColumns[ii], (Comparable<?>) partitionData.get(ii));
```

Comment:
I wonder if we need some kind of type coercion here. We're throwing some opaque object into the map and hoping it maps to the expected Java class when we pull it out later.

Reply:
Can we validate that the value matches the type of the associated column from the table definition?